pkg/tc/util.go

```diff
@@ -0,0 +1,321 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package trafficctl
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/florianl/go-tc"
+)
+
+const (
+	TimeUnitsPerSecs = 1000000
+)
+
+func parseTime(t string) (uint32, error) {
+	var cutstring string
+	multiplier := TimeUnitsPerSecs
+	if strings.HasSuffix(t, "sec") {
+		cutstring, _ = strings.CutSuffix(t, "sec")
+		multiplier = TimeUnitsPerSecs
+	} else if strings.HasSuffix(t, "secs") {
+		cutstring, _ = strings.CutSuffix(t, "secs")
+		multiplier = TimeUnitsPerSecs
+	} else if strings.HasSuffix(t, "s") {
+		cutstring, _ = strings.CutSuffix(t, "s")
+		multiplier = TimeUnitsPerSecs
+	}
+
+	if strings.HasSuffix(t, "ms") {
+		cutstring, _ = strings.CutSuffix(t, "ms")
+		multiplier = TimeUnitsPerSecs / 1000
+	} else if strings.HasSuffix(t, "msec") {
+		cutstring, _ = strings.CutSuffix(t, "msec")
+		multiplier = TimeUnitsPerSecs / 1000
+	} else if strings.HasSuffix(t, "msecs") {
+		cutstring, _ = strings.CutSuffix(t, "msecs")
+		multiplier = TimeUnitsPerSecs / 1000
+	}
+
+	if strings.HasSuffix(t, "us") {
+		cutstring, _ = strings.CutSuffix(t, "us")
+		multiplier = TimeUnitsPerSecs / 1000000
+	} else if strings.HasSuffix(t, "usec") {
+		cutstring, _ = strings.CutSuffix(t, "usec")
+		multiplier = TimeUnitsPerSecs / 1000000
+	} else if strings.HasSuffix(t, "usecs") {
+		cutstring, _ = strings.CutSuffix(t, "usecs")
+		multiplier = TimeUnitsPerSecs / 1000000
+	}
+
+	val, err := strconv.Atoi(cutstring)
+	if err != nil {
+		return 0, err
+	}
+	if val < 0x0 || val >= 0x7FFFFFFF {
+		return 0, ErrOutOfBounds
+	}
+
+	ret := uint32(val) * uint32(multiplier)
+
+	return ret, nil
+}
+
+func ParseHandle(h string) (uint32, error) {
+	// split the string at :
+	maj, _, ok := strings.Cut(h, ":")
+	if !ok {
+		return 0, ErrInvalidArg
+	}
+
+	major, err := strconv.ParseUint(maj, 16, 16)
+	if err != nil {
+		return 0, err
+	}
+
+	return uint32(major) << 16, nil
+}
+
+func ParseClassID(p string) (uint32, error) {
+	if p == "root" {
+		return tc.HandleRoot, nil
+	}
+
+	if p == "none" {
+		return 0, nil
+	}
+
+	// split the string at :
+	maj, min, ok := strings.Cut(p, ":")
+	if !ok {
+		return 0, ErrInvalidArg
+	}
+
+	major, err := strconv.ParseUint(maj, 16, 16)
+	if err != nil {
+		return 0, err
+	}
+
+	if min == "" {
+		return uint32(major << 16), nil
+	}
+	minor, err := strconv.ParseUint(min, 16, 16)
+	if err != nil {
+		return 0, err
+	}
+
+	return uint32(major<<16) | uint32(minor), nil
+}
+
+var (
+	ErrUnknownLinkLayer = errors.New("unknown linklayer value provided")
+)
+
+func ParseLinkLayer(l string) (uint8, error) {
+	for _, ll := range []struct {
+		name string
+		val  uint8
+	}{
+		{name: "ethernet", val: 1},
+		{name: "atm", val: 2},
+		{name: "ads1", val: 2},
+	} {
+		if ll.name == l {
+			return ll.val, nil
+		}
+	}
+	return 0xFF, ErrUnknownLinkLayer
+}
+
+func ParseSize(s string) (uint64, error) {
```

Comment:
does ulikunits have this already? I can't believe this is not already in u-root somewhere.

Reply:
Cant find such thing ;(